File: scripts/python/test_scheduler.py
from NSCP import Settings, Registry, Core, log, status, log_debug, log_error, sleep
from test_helper import BasicTest, TestResult, Callable, setup_singleton, install_testcases, init_testcases, shutdown_testcases
from types import *
from time import time
import random
import os

prefix = 'scheduler'

class SchedulerTest(BasicTest):

    check_count = 0
    results_count = 0
    command_count = {}
    
    sched_alias = 'test_sched_%s'%prefix
    python_channel = 'test_sched_%s_py'%prefix
    command = 'test_sched_%s'%prefix
    sched_base_path = '/settings/%s'%sched_alias
    
    def simple_check_handler(arguments):
        instance = SchedulerTest.getInstance()
        return instance.wrapped_simple_check_handler(arguments)
    simple_check_handler = Callable(simple_check_handler)

    def wrapped_simple_check_handler(self, arguments):
        self.check_count = self.check_count + 1
        if arguments:
            if not arguments[0] in self.command_count:
                self.command_count[arguments[0]] = 1
            else:
                self.command_count[arguments[0]] = self.command_count[arguments[0]] + 1
            return (status.OK, arguments[0], '')
        return (status.OK, 'pong', '')

    def on_stress_handler(channel, source, command, code, message, perf):
        instance = SchedulerTest.getInstance()
        instance.wrapped_on_stress_handler(channel, source, command, code, message, perf)
    on_stress_handler = Callable(on_stress_handler)

    def wrapped_on_stress_handler(self, channel, source, command, code, message, perf):
        self.results_count = self.results_count + 1
        return None

    def desc(self):
        return 'Testcase for Scheduler'

    def title(self):
        return 'Test Scheduler'

    def setup(self, plugin_id, prefix):
        self.reg.simple_function(self.command, SchedulerTest.simple_check_handler, 'This is a simple noop command')
        self.reg.simple_subscription(self.python_channel, SchedulerTest.on_stress_handler)

        #self.core.reload('%s,delayed'%self.sched_alias)
        

    def teardown(self):
        self.conf.set_string(self.sched_base_path, 'threads', '0')
        self.core.reload(self.sched_alias)

    def check_one(self, result, key, min, max):
        result.assert_gt(self.command_count[key], min, 'check %s (%d) fired more then %d'%(key, self.command_count[key], min))
        result.assert_lt(self.command_count[key], max, 'check %s (%d) fired less then %d'%(key, self.command_count[key], max))
    
    def run_test(self):
        self.core.load_module('Scheduler', self.sched_alias)
        result = TestResult()
        start = time()

        last_major = 0
        elapsed = time()-start
        while elapsed < 60:
            if elapsed > 0:
                log("testing scheduler %d%% (collected %d instance in %d seconds)"%(elapsed/60*100, self.results_count, elapsed))
            sleep(2000)
            elapsed = time()-start
        result.add_message(True, 'Summary Collected %d instance in %d seconds: %d/s'%(self.results_count, elapsed, self.results_count/elapsed))
        self.check_one(result, "rand", 5, 10)
        self.check_one(result, "1s", 55, 65)
        self.check_one(result, "short", 10, 14)
        self.check_one(result, "30s", 1, 3)
        self.check_one(result, "explicit", 10, 14)
        self.check_one(result, "10s", 5, 7)

        return result

    def install(self, arguments):
        # Configure required modules
        self.conf.set_string('/modules', 'pytest', 'PythonScript')
        #self.conf.set_string('/modules', self.sched_alias, 'Scheduler')
        self.conf.set_string('/modules', 'CheckSystem', 'enabled')

        # Configure python
        self.conf.set_string('/settings/pytest/scripts', 'test_stress', 'test_scheduler.py')
        
        default_path = '%s/schedules/default'%self.sched_base_path
        self.conf.set_string(default_path, 'channel', self.python_channel)
        self.conf.set_string(default_path, 'command', "%s default"%self.command)
        self.conf.set_string(default_path, 'interval', '5s')
        self.conf.set_string(default_path, 'randomness', '0%')

        self.conf.set_string('%s/schedules'%(self.sched_base_path), 'python_checker_d', "%s short"%self.command)

        self.conf.set_string('%s/schedules/python_checker_e'%(self.sched_base_path), 'command', "%s explicit"%self.command)

        #self.conf.set_string('%s/schedules/python_checker_i'%(self.sched_base_path), 'interval', '1s')

        self.conf.set_string('%s/schedules/python_checker_1s'%(self.sched_base_path), 'command', "%s 1s"%self.command)
        self.conf.set_string('%s/schedules/python_checker_1s'%(self.sched_base_path), 'interval', '1s')

        self.conf.set_string('%s/schedules/python_checker_10s'%(self.sched_base_path), 'command', "%s 10s"%self.command)
        self.conf.set_string('%s/schedules/python_checker_10s'%(self.sched_base_path), 'interval', '10s')

        self.conf.set_string('%s/schedules/python_checker_30s'%(self.sched_base_path), 'command', "%s 30s"%self.command)
        self.conf.set_string('%s/schedules/python_checker_30s'%(self.sched_base_path), 'interval', '30s')

        self.conf.set_string('%s/schedules/python_checker_r10s'%(self.sched_base_path), 'command', "%s rand"%self.command)
        self.conf.set_string('%s/schedules/python_checker_r10s'%(self.sched_base_path), 'interval', '10s')
        self.conf.set_string('%s/schedules/python_checker_r10s'%(self.sched_base_path), 'randomness', '50%')

        self.conf.save()

    def uninstall(self):
        None

    def help(self):
        None

    def init(self, plugin_id, prefix):
        self.reg = Registry.get(plugin_id)
        self.conf = Settings.get(plugin_id)
        self.core = Core.get(plugin_id)
        None

    def shutdown(self):
        None

    def require_boot(self):
        return True

setup_singleton(SchedulerTest)

all_tests = [SchedulerTest]

def __main__(args):
    install_testcases(all_tests)
    
def init(plugin_id, plugin_alias, script_alias):
    init_testcases(plugin_id, plugin_alias, script_alias, all_tests)

def shutdown():
    shutdown_testcases()

File: scripts/lua/test_scheduler.lua
-- Protobuf-free Lua port of scripts/python/test_scheduler.py.
--
-- Verifies the Scheduler module actually fires scheduled checks at their
-- configured intervals and routes the results through a submission channel:
--   * a Lua simple_function (test_sched_cmd) is the scheduled check - it counts
--     how often it runs, keyed by its argument;
--   * a Lua simple_subscription on a channel counts the results the scheduler
--     submits back.
-- After a sampling window we assert each interval fired within a plausible band
-- and that faster intervals fired more often than slower ones. This drives the
-- Scheduler worker threads + core submit path, a good sanitizer target. No
-- protobuf, no network.
--
-- The Python original samples for 90s to tightly bound every interval; this port
-- uses a shorter window with looser bounds (plus the robust "faster fires more"
-- invariant) so it stays reliable as a CI/sanitizer test.
local test = require("test_helper")

local ALIAS = 'test_sched'              -- Scheduler module instance + settings root
local BASE = '/settings/' .. ALIAS
local CHANNEL = 'test_sched_py'         -- channel the scheduler submits results to
local COMMAND = 'test_sched_cmd'        -- the scheduled (Lua) check command

-- Sampling window in seconds; the per-interval bounds below are calibrated for it.
local WINDOW = 16

-- Shared counters (single Lua state, so plain upvalues are fine).
local check_count = 0
local results_count = 0
local command_count = {}

-- The scheduled check: count invocations keyed by the first argument (which the
-- schedules below set to the interval name, e.g. "1s"), and echo it back.
local function check_handler(command, args)
	check_count = check_count + 1
	local key = args[1]
	if key then
		command_count[key] = (command_count[key] or 0) + 1
		return 'ok', key, ''
	end
	return 'ok', 'pong', ''
end

-- The result sink: the scheduler submits each (OK) check result here.
-- Signature/return match the simple_subscription contract (channel, command,
-- result, lines) -> (status, message).
local function on_result(channel, command, result, lines)
	results_count = results_count + 1
	return 'ok', 'received'
end

local reg = Registry()
reg:simple_function(COMMAND, check_handler, 'noop scheduler check')
reg:simple_subscription(CHANNEL, on_result, 'scheduler result sink')

local SchedulerTest = {}

function SchedulerTest:install(arguments)
	local conf = Settings()
	conf:set_string('/modules', ALIAS, 'Scheduler')
	conf:set_string('/modules', 'luatest', 'LUAScript')
	conf:set_string('/settings/luatest/scripts', 'test_scheduler', 'test_scheduler.lua')

	-- Default schedule: provides the channel inherited by the others, plus its
	-- own 5s beat.
	conf:set_string(BASE .. '/schedules/default', 'channel', CHANNEL)
	conf:set_string(BASE .. '/schedules/default', 'command', COMMAND .. ' default')
	conf:set_string(BASE .. '/schedules/default', 'interval', '5s')
	conf:set_string(BASE .. '/schedules/default', 'randomness', '0%')

	conf:set_string(BASE .. '/schedules/checker_1s', 'command', COMMAND .. ' 1s')
	conf:set_string(BASE .. '/schedules/checker_1s', 'interval', '1s')

	conf:set_string(BASE .. '/schedules/checker_5s', 'command', COMMAND .. ' 5s')
	conf:set_string(BASE .. '/schedules/checker_5s', 'interval', '5s')

	conf:set_string(BASE .. '/schedules/checker_10s', 'command', COMMAND .. ' 10s')
	conf:set_string(BASE .. '/schedules/checker_10s', 'interval', '10s')
end

function SchedulerTest:setup() end

function SchedulerTest:teardown()
	-- Stop the scheduler so it does not keep firing after the test.
	local conf = Settings()
	conf:set_string(BASE, 'threads', '0')
	Core():reload(ALIAS)
end

-- Assert command `key` fired strictly between min and max times in the window.
function SchedulerTest:check_one(result, key, min, max)
	local n = command_count[key] or 0
	result:add_message(n > min and n < max,
		string.format('check %s fired %d times (expected %d..%d)', key, n, min, max))
end

function SchedulerTest:run()
	local core = Core()
	local result = test.TestResult:new{message = 'Scheduler fires checks and routes results'}

	-- The scheduler has been running since boot; reset and measure a clean window.
	check_count = 0
	results_count = 0
	command_count = {}

	local start = os.time()
	while os.time() - start < WINDOW do
		nscp.sleep(2000)
		local elapsed = os.time() - start
		core:log('info', string.format('scheduler test: %ds, %d results so far', elapsed, results_count))
	end

	result:add_message(check_count > 0, string.format('scheduled checks fired (%d)', check_count))
	result:add_message(results_count > 0, string.format('results delivered to channel (%d)', results_count))

	-- Expected counts ~= WINDOW / interval; bounds absorb missed/extra beats.
	--   1s -> ~16, 5s (and default) -> ~3, 10s -> ~1.
	-- ('default' is the inheritance template, not an executed schedule, so it
	-- never fires under its own key - we don't assert it.)
	self:check_one(result, '1s', 8, 30)
	self:check_one(result, '5s', 1, 8)
	self:check_one(result, '10s', 0, 5)

	-- Robust ordering invariant: faster intervals must fire more often.
	result:add_message((command_count['1s'] or 0) > (command_count['5s'] or 0),
		'1s interval fired more often than 5s')
	result:add_message((command_count['5s'] or 0) >= (command_count['10s'] or 0),
		'5s interval fired at least as often as 10s')

	return result
end

local instances = { SchedulerTest }
test.init_test_manager(instances)

function main(args)
	return test.install_test_manager(instances)
end
